### PR TITLE
Add back latest arbitrator factory registry code

### PR DESCRIPTION
**Pull request by zhztheplayer**
The previous [PR](https://github.com/facebookincubator/velox/pull/5965) landed wrong version due to Meta
internal import tool problem.

### CODE REVIEW

**Comment by zhztheplayer** [comment]
@xiaoxmeng I've updated the PR.

File: velox/exec/tests/utils/OperatorTestBase.cpp [hunk 1]
```diff
@@ -57,6 +58,9 @@ void OperatorTestBase::TearDownTestCase() {
 }
 
 void OperatorTestBase::SetUp() {
+  if (!memory::SharedArbitrator::isFactoryRegistered()) {
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply velox/exec/tests/utils/OperatorTestBase.cpp:61]: That will fail some test cases that are executed in the same process.

File: velox/common/memory/MemoryArbitrator.h [hunk 2]
```diff
@@ -91,12 +68,29 @@ class MemoryArbitrator {
...
+
+  /// Check if factory is registered for the kind.
+  static bool isFactoryRegistered(const std::string& kind);
+  /// Invoked by the memory manager to create an instance of memory arbitrator
+  /// based on the kind specified in 'config'. The arbitrator kind must be
+  /// registered through MemoryArbitrator::registerFactory, otherwise the
+  /// function throws a velox user exception error.
+  ///
+  /// NOTE: if arbitrator kind is not set in 'config', then the function returns
+  /// nullptr.
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply velox/common/memory/MemoryArbitrator.h:94]: done

File: velox/common/memory/MemoryArbitrator.h [hunk 3]
```diff
@@ -91,12 +68,29 @@ class MemoryArbitrator {
...
+  using Factory = std::function<std::unique_ptr<MemoryArbitrator>(
+      const MemoryArbitrator::Config& config)>;
+
+  /// Register factory for a specific kind of memory arbitrator.
+  /// MemoryArbitrator::Create looks up the registry to find the factory to
+  /// create arbitrator instance based on the kind specified in arbitrator
+  /// config.
+  static void registerFactory(const std::string& kind, Factory factory);
+
+  /// Check if factory is registered for the kind.
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply velox/common/memory/MemoryArbitrator.h:85]: "isRegistered" is somehow necessary if we had a "already registered" check in "register(kind, ..)".

I looked into Velox's code then noticed that It's a common practice not having a "already registered" check in "register" routines. That's why only this patch requires for "isRegistered". However I am inclined that the way in this PR is slightly better (safer, immutable) than the current ways. 

However I can still make changes if you insist, it's not something critical to the overall functionality.

File: velox/common/testutil/TestValue.h [hunk 4]
```diff
@@ -47,8 +47,8 @@ class TestValue {
       const std::string& injectionPoint,
       std::function<void(T*)> injectionCb);
 
-  /// Invoked by the test code to unregister a callback hook at the specified
-  /// execution point.
+  /// Invoked by the test code to unregisterFactory a callback hook at the
```
Comment (xiaoxmeng) [inline velox/common/testutil/TestValue.h:50]: Can fix the comment here?

**Review by xiaoxmeng — COMMENTED** [review]
@zhztheplayer thanks for the iterating!

**Pull request closed without merging**